semgrep-core/Analyzing/Bloom_annotation.ml

```diff
@@ -53,23 +91,94 @@ module B = Bloom_filter
  * See also Rules_filter.reserved_id and reserved_str and
  * Rules_filter.extract_specific_strings.
 *)
-let extract_strings _any =
-  []
+
+let rec statement_strings stmt =
+  let res = ref [] in
+  let top_level = ref true in
+  let visitor = V.mk_visitor {
+    V.default_visitor with
+    V.kident = (fun (_k, _) (str, _tok) ->
+      if not (special_ident str) then
+        push str res
+    );
+    V.kexpr = (fun (k, _) x ->
+      (match x with
+       (* less: we could extract strings for the other literals too?
+        * atoms, chars, even int?
+       *)
+       | L (String (str, _tok)) ->
+           if not (special_literal str) then push str res
+       | TypedMetavar _ -> ()
+       | _ -> k x
+      )
+    );
+    V.kconstness = (fun (k, _) x ->
+      (match x with
+       | Lit (String (str, _tok)) ->
+           if not (special_literal str) then push str res
+       | _ -> k x
+      )
+    );
+    V.kid_info = (fun (k, _) x -> k x);
+    V.kstmt = (fun (k, _) x ->
+      (* First statement visited is the current statement *)
+      if !top_level then (top_level := false; k x)
+      else
+        (* For any other statement, recurse to add the filter *)
+        (let strs = statement_strings x in
+         let bf = B.create () in
+         add_all_to_bloom strs bf;
+         push_list strs res;
+         x.s_bf <- Some (bf))
+    );
+  } in
+  visitor (S stmt);
+  !res
+;;
 
 (*****************************************************************************)
 (* Analyze the pattern *)
 (*****************************************************************************)
 
-let bloom_of_expr _e =
-  B.create ()
-
-let bloom_of_stmt _st =
-  B.create ()
+let list_of_pattern_strings any =
+  let res = ref [] in
+  let visitor = V.mk_visitor {
+    V.default_visitor with
```

Comment:
looks like duplicated code with above. Maybe you can factorize.

Reply:
I wasn't sure what the best way to do this was. The code is very similar in some ways, but statement_strings is a recursive function, called recursively at each statement node, whereas list_of_pattern_strings visits the entire AST. I could make functions to produce kident and kexpr which take in a ref, I just wasn't sure if it would end up making the code less readable. If I start making kexpr more complicated, I'll factorize it.

Reply:
ok, sounds fair.